exams/pearson/audit.py

```diff
@@ -0,0 +1,265 @@
+import gnupg
+
+SSE_ENCRYPTION_ALGORITHM = 'AES256'
+REQUIRED_SETTINGS = [
+    'EXAMS_AUDIT_ENCRYPTION_FINGERPRINT',
+    'EXAMS_AUDIT_ENCRYPTION_PUBLIC_KEY',
+]
+
+log = logging.getLogger(__name__)
+
+
+class S3AuditStorage:
+    """Audit storage mechanism for S3"""
+
+    def get_connection(self):
+        """
+        Creates a connection to S3
+
+        Raises:
+            ImproperlyConfigured
+                if a required setting is not set
+        Returns:
+            boto.s3.connection.S3Connection
+                a connection to S3
+        """
+        if any([key is None for key in [
+                settings.EXAMS_AUDIT_AWS_ACCESS_KEY_ID,
+                settings.EXAMS_AUDIT_AWS_SECRET_ACCESS_KEY,
+        ]]):
+            raise ImproperlyConfigured(
+                "One of EXAMS_AUDIT_AWS_ACCESS_KEY_ID or EXAMS_AUDIT_AWS_SECRET_ACCESS_KEY is required but not set"
+            )
+        return connection.S3Connection(
+            settings.EXAMS_AUDIT_AWS_ACCESS_KEY_ID,
+            settings.EXAMS_AUDIT_AWS_SECRET_ACCESS_KEY
+        )
+
+    def get_bucket(self):
+        """
+        Gets the configured bucket
+
+        Raises:
+            ImproperlyConfigured
+                if a required setting is not set
+        Returns:
+            boto.s3.bucket.Bucket
+                the S3 bucket for storage
+        """
+        if settings.EXAMS_AUDIT_S3_BUCKET is None:
+            raise ImproperlyConfigured('Setting EXAMS_AUDIT_S3_BUCKET is required but not set')
+        return self.get_connection().get_bucket(settings.EXAMS_AUDIT_S3_BUCKET)
+
+    def get_s3_key(self, filename, file_type):
+        """
+        Determines the S3 key to store the file under
+
+        Args:
+            filename (str):
+                filename of the encrypted file
+            file_type (str):
+                type of the encrypted file
+
+        Returns:
+            boto.s3.key.Key:
+                the key to store the file in
+        """
+        _, filename = os.path.split(filename)
+        return key.Key(
+            self.get_bucket(),
+            'exam_audits/{file_type}/{filename}'.format(
+                filename=filename,
+                file_type=file_type,
+            )
+        )
+
+    def upload(self, filename, file_type):
+        """
+        Determines the S3 key to store the file under
+
+        Args:
+            filename (str):
+                filename of the encrypted file
+            file_type (str):
+                type of the encrypted file
+
+        Returns:
+            str:
+                the key path in S3 where the file way stored
+        """
+        s3_key = self.get_s3_key(filename, file_type)
+        s3_key.set_contents_from_filename(
+            filename,
+            headers={
+                'x-amz-server-side-encryption': SSE_ENCRYPTION_ALGORITHM,
+            }
+        )
+        return s3_key.key
+
+
+class ExamDataAuditor:
```

Comment:
Using a class was intentional, although you're pointing out one oversight I forgot to address - I want to only configure this class and cache that state in memory because we could be potentially auditing a lot of files in one run and I'd prefer not to keep performing IO to inspect state that shouldn't change under normal circumstances. I need to store that state somewhere so a class is best for that.